### ISSUE TITLE
Implement Key polynomials

### ISSUE THREAD

**Issue by trevorkarn** [sagemath/sage#34414]
<div id="comment:0"></div>

Implement the Key polynomials as an algebra with basis indexed by compositions.

Depends on #34435
Depends on #34510
Depends on #34527
Depends on #34535
Depends on #34581

CC:  @trevorkarn @tscrim

Component: **combinatorics**

Keywords: **gsoc2022 key-polynomial**

Author: **Trevor K. Karn**

Branch/Commit: **[u/tscrim/key_polynomials-34414](https://github.com/sagemath/sagetrac-mirror/tree/u/tscrim/key_polynomials-34414) @ [`124900f`](https://github.com/sagemath/sagetrac-mirror/commit/124900faa1c5ce221dd3076789a0d4b64d6376d1)**

Reviewer: **Travis Scrimshaw**

_Issue created by migration from https://trac.sagemath.org/ticket/34414_

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`3805067`](https://github.com/sagemath/sagetrac-mirror/commit/3805067c50bdd7a3e24d20221c022b9dd15e2427)** to **[`a96d274`](https://github.com/sagemath/sagetrac-mirror/commit/a96d2749ad11d4cbf42bc7d006cd7fda31c10518)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:38"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/a96d2749ad11d4cbf42bc7d006cd7fda31c10518"><code>a96d274</code></a></td><td><code>Clarify sorting word documentation</code></td></tr></table>

**Comment by trevorkarn** [comment]
<div id="comment:39" align="right">comment:39</div>

Is there anything I am forgetting to do on this?

**Comment by tscrim** [comment]
<div id="comment:40" align="right">comment:40</div>

Nothing you are forgetting I think, but you can simplify the s<sub>i</sub> action:

```python
f = # some polynomial
g = f.parent().gens()
sf = f.subs({g[i-1]:g[i], g[i]:g[i-1]})
```
I think works (I cannot test right now). Or `f(**{repr(g[i-1]):g[i], repr(g[i]): g[i-1]})` should work.

**Comment by trevorkarn** [comment]
<div id="comment:41" align="right">comment:41</div>

Replying to [Travis Scrimshaw](#comment%3A40):
> Nothing you are forgetting I think, but you can simplify the s<sub>i</sub> action:
> 
> ```python
> f = # some polynomial
> g = f.parent().gens()
> sf = f.subs({g[i-1]:g[i], g[i]:g[i-1]})
> ```
> I think works (I cannot test right now). Or `f(**{repr(g[i-1]):g[i], repr(g[i]): g[i-1]})` should work.


Turns out this only works in the finite case. In the infinite case:

```
sage: R.<z> = InfinitePolynomialRing(QQ)
sage: f = z[1] + z[1]*z[2]*z[3]
sage: f.subs({z[1]:z[0]})
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [3], in <cell line: 1>()
----> 1 f.subs({z[Integer(1)]:z[Integer(0)]})

File ~/Applications/sage/src/sage/structure/element.pyx:830, in sage.structure.element.Element.subs()
    828 if str(gen) in kwds:
    829     variables.append(kwds[str(gen)])
--> 830 elif in_dict and gen in in_dict:
    831     variables.append(in_dict[gen])
    832 else:

TypeError: unhashable type: 'InfinitePolynomialGen'
```

See #34581.

**Comment by trevorkarn** [comment]
Changed dependencies from **#34435 #34510 #34527 #34535** to **#34435 #34510 #34527 #34535 #34581**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`a96d274`](https://github.com/sagemath/sagetrac-mirror/commit/a96d2749ad11d4cbf42bc7d006cd7fda31c10518)** to **[`c32112a`](https://github.com/sagemath/sagetrac-mirror/commit/c32112a26bbea68ac04be7ce93112605cfef3e4e)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:43"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/fbf32283d56b5dde7b6071b7fadcec25dbe77e13"><code>fbf3228</code></a></td><td><code>Simplify subs in finite case</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/c32112a26bbea68ac04be7ce93112605cfef3e4e"><code>c32112a</code></a></td><td><code>Simplify action on infinite polynomial ring</code></td></tr></table>

**Comment by trevorkarn** [comment]
<div id="comment:44" align="right">comment:44</div>

Replying to [Travis Scrimshaw](#comment%3A40):
> you can simplify the s<sub>i</sub> action

Done in both cases without using `.subs()` for the infinite case.

**Comment by trevorkarn** [comment]
Changed dependencies from **#34435 #34510 #34527 #34535 #34581** to **#34435 #34510 #34527 #34535**

**Comment by tscrim** [comment]
<div id="comment:46" align="right">comment:46</div>

Thanks for the updates.

I don't quite understand why that code works for the infinite case. I think depending on #34581 is the right way to go. We should be able to fix that one fairly quickly.

Some more comments on this:

- Rename `poly_gen()` -> `poly_gens()` as it returns all of the generators.
- `while f.monomials():` seems wasteful; you can simply check `while f:` (is not 0).
- In `from_polynomial()`, I would separate out the `self._indices(m).trim()` into a separate helper method

  ```python
  if self._k:
      def build_index(m):
          return self._indices(m)
  else:
      def build_index(m):
          return self._indices(m).trim()
  ```
  Furthermore I am not certain that the monomials and exponents will always be given in lex order. While it is a bit more computationally intensive, I think you should call `max(f.exponents())`. You can then use that to build the monomial and/or get the coefficient. This allows you to remove the code duplication.
- Since you have an element class, you are better implementing `_mul_` there as that avoids some indirection.

**Comment by tscrim** [comment]
Changed dependencies from **#34435 #34510 #34527 #34535** to **#34435 #34510 #34527 #34535 #34581**

**Comment by tscrim** [comment]
Reviewer: **Travis Scrimshaw**

**Comment by trevorkarn** [comment]
<div id="comment:47" align="right">comment:47</div>

Also try when the base ring is something more interesting than `ZZ` or `QQ`.

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`c32112a`](https://github.com/sagemath/sagetrac-mirror/commit/c32112a26bbea68ac04be7ce93112605cfef3e4e)** to **[`cd7d6b5`](https://github.com/sagemath/sagetrac-mirror/commit/cd7d6b5fae2968363a21e65b40f79891973bacb4)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:48"></div>

Branch pushed to git repo; I updated commit sha1. This was a forced push. Last 10 new commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/58e6f17cf45414e2be71c8ba17e8bfbcd564b8db"><code>58e6f17</code></a></td><td><code>Fix vector length bug</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/0bbaf78ad10b14c1e1558b2213ef14217384657f"><code>0bbaf78</code></a></td><td><code>Add tests</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/fdd2407d2b17acf3e07c422f86447e01d0ecf9a1"><code>fdd2407</code></a></td><td><code>self._basis_keys -> self._indices, simplify key polynomial basis repr, add a sentence for class level description</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/3bc2793945193ce5112878a06951d2264994cd76"><code>3bc2793</code></a></td><td><code>Move divided difference away from monomial</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/6f8ea0279fe7bed4c7b9a523ccb2c33304baf9c6"><code>6f8ea02</code></a></td><td><code>Sequence -> Collection</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/c756241fa874e36a1e978c8826f236b57f3cb391"><code>c756241</code></a></td><td><code>Simplify division</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/4153a9e876cb8b0af904e7a5013b297fc8d143d8"><code>4153a9e</code></a></td><td><code>Clarify sorting word documentation</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/40e135ec0f1ffc215afbd3e3d49d2aaed4ee5eb6"><code>40e135e</code></a></td><td><code>Simplify subs in finite case</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/e7313bc060fbe0cda30f2e3a2126fcd1f3ab10ec"><code>e7313bc</code></a></td><td><code>Simplify action on infinite polynomial ring</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/cd7d6b5fae2968363a21e65b40f79891973bacb4"><code>cd7d6b5</code></a></td><td><code>poly_gen -> poly_gens, uniformly use .subs, and while f.monomials -> while f</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:49"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/d2776543f465a77be82901da1ecf9bbf34e64a02"><code>d277654</code></a></td><td><code>Remove duplicated code</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/0ab036ad218986914f1a67af53a2819fb9fcf36d"><code>0ab036a</code></a></td><td><code>Add `_mul_` and remove .product()</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/70e46873bdd5ec0c4d2a5cd4e1b9e2b41c8fa246"><code>70e4687</code></a></td><td><code>Add test for rational division and try with coefficients in GF(5)</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`cd7d6b5`](https://github.com/sagemath/sagetrac-mirror/commit/cd7d6b5fae2968363a21e65b40f79891973bacb4)** to **[`70e4687`](https://github.com/sagemath/sagetrac-mirror/commit/70e46873bdd5ec0c4d2a5cd4e1b9e2b41c8fa246)**

**Comment by trevorkarn** [comment]
<div id="comment:50" align="right">comment:50</div>

Replying to [Travis Scrimshaw](#comment%3A46):
> I don't quite understand why that code works for the infinite case. I think depending on #34581 is the right way to go. We should be able to fix that one fairly quickly.
> 
> Some more comments on this:
> 
> - Rename `poly_gen()` -> `poly_gens()` as it returns all of the generators.
> - `while f.monomials():` seems wasteful; you can simply check `while f:` (is not 0).
> - In `from_polynomial()`, I would separate out the `self._indices(m).trim()` into a separate helper method
> 
>   ```python
>   if self._k:
>       def build_index(m):
>           return self._indices(m)
>   else:
>       def build_index(m):
>           return self._indices(m).trim()
>   ```

Done

>   Furthermore I am not certain that the monomials and exponents will always be given in lex order. While it is a bit more computationally intensive, I think you should call `max(f.exponents())`. You can then use that to build the monomial and/or get the coefficient. This allows you to remove the code duplication.

I avoided this by picking the monomial and then getting the exponent from the original polynomial instead getting it from the monomial itself. I think this shores up the weak point in my previous approach.

> - Since you have an element class, you are better implementing `_mul_` there as that avoids some indirection.

Done.

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`70e4687`](https://github.com/sagemath/sagetrac-mirror/commit/70e46873bdd5ec0c4d2a5cd4e1b9e2b41c8fa246)** to **[`197b53e`](https://github.com/sagemath/sagetrac-mirror/commit/197b53e3db4adc8620a1fb9cbcacd64341a42153)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:52"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/197b53e3db4adc8620a1fb9cbcacd64341a42153"><code>197b53e</code></a></td><td><code>PEP8 compliance</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`197b53e`](https://github.com/sagemath/sagetrac-mirror/commit/197b53e3db4adc8620a1fb9cbcacd64341a42153)** to **[`4855ece`](https://github.com/sagemath/sagetrac-mirror/commit/4855ece48594a47c3d565477da7602b44592df2e)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:53"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/4855ece48594a47c3d565477da7602b44592df2e"><code>4855ece</code></a></td><td><code>Fix pyflakes issues</code></td></tr></table>

**Comment by tscrim** [comment]
Changed branch from **[u/tkarn/key-poly-34414](https://github.com/sagemath/sagetrac-mirror/tree/u/tkarn/key-poly-34414)** to **[u/tscrim/key_polynomials-34414](https://github.com/sagemath/sagetrac-mirror/tree/u/tscrim/key_polynomials-34414)**

**Comment by tscrim** [comment]
Changed commit from **[`4855ece`](https://github.com/sagemath/sagetrac-mirror/commit/4855ece48594a47c3d565477da7602b44592df2e)** to **[`2e2c4f2`](https://github.com/sagemath/sagetrac-mirror/commit/2e2c4f2990070e8f04bd5ace80aa4570eceeb54c)**

**Comment by tscrim** [comment]
<div id="comment:54" align="right">comment:54</div>

I (finally) got around to doing the review changes I wanted. The biggest change is I made everything 1-based, which will avoid a lot of confusion from users. I also made the divided difference operators more usable by general polynomials and minimized some of the function calls (at the expense of some 1-line code duplication). If my changes are good, then this can be set to a positive review.

---
New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/710cfbbd6cefa381183746fbff6fd99ec46b6891"><code>710cfbb</code></a></td><td><code>Reviewer changes: refactored some of the divided difference code, doc changes, orther misc changes.</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/5e8977fe18f6b38e9eb4aeca066438b8d7e0514d"><code>5e8977f</code></a></td><td><code>Add reviewer suggestions to docs and change behavior of conflicting subs</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/a56e345cd9c3203b432f2f29c743f0f4606dc088"><code>a56e345</code></a></td><td><code>Doc change</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/2e2c4f2990070e8f04bd5ace80aa4570eceeb54c"><code>2e2c4f2</code></a></td><td><code>Merge branch 'u/tkarn/inf-poly-subs-34581' of https://github.com/sagemath/sagetrac-mirror into u/tscrim/key_polynomials-34414</code></td></tr></table>

**Comment by trevorkarn** [comment]
<div id="comment:55" align="right">comment:55</div>

Thanks! These all look good to me. Especially implementing `__getitem__`. Setting to positive review.

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`2e2c4f2`](https://github.com/sagemath/sagetrac-mirror/commit/2e2c4f2990070e8f04bd5ace80aa4570eceeb54c)** to **[`124900f`](https://github.com/sagemath/sagetrac-mirror/commit/124900faa1c5ce221dd3076789a0d4b64d6376d1)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:57"></div>

Branch pushed to git repo; I updated commit sha1 and set ticket back to needs_review. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/34bbd610cf79e13bf291e2dd35de215f70b50789"><code>34bbd61</code></a></td><td><code>Merge branch 'u/tscrim/key_polynomials-34414' of https://github.com/sagemath/sagetrac-mirror into u/tscrim/key_polynomials-34414</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/124900faa1c5ce221dd3076789a0d4b64d6376d1"><code>124900f</code></a></td><td><code>Removed unused copy for pyflakes.</code></td></tr></table>

**Comment by tscrim** [comment]
<div id="comment:58" align="right">comment:58</div>

Trivial change to make pyflakes happy.

**Comment by mkoeppe** [comment]
Merged in 10.0.beta0
